packages/dataviews/src/types.ts

```diff
@@ -1,7 +1,9 @@
 /**
  * External dependencies
  */
-import type { ReactElement, ReactNode } from 'react';
+import type { ReactElement, ReactNode, Dispatch, SetStateAction } from 'react';
+
+export type setSelection = Dispatch< SetStateAction< string[] > >;
```

Comment:
What does "dispatch" do?

--
The rest of the comment here is not blocking, mostly discussion and typescript learnings / guidelines discussion.

I think there's a balance to be found between the types that make sense as exported or not. For smaller types like this, I think duplication may be fine (avoid indirection). But I don't know if folks more familiar with typescript think. cc @sirreal @jsnajdr 

Also curious what @mcsf thinks here.

Reply:
I can revert 2e2f38c4764242cbcb739edfc083901eea616e3f if you like

Reply:
Oh I see what Dispatch does, it's useful I'd say :) 

Curious about the guidelines on exposing this kind of low level types.

Reply:
Yeah, I forgot I changed a public type too. I can revert it to `( selection: string[] | ( ( prevState: string[] ) => string[] ) ) => void` to unblock if you prefer.

Reply:
My feedback is that it just took me non-trivial effort to figure out what this `Dispatch<SetStateAction>` actually is. Better to write it plainly as @ellatrix proposes. Also make the first letter of the type name uppercase, `SetSelection` 🙂

Reply:
Convention is typically to capitalize types, it might be good to change the type name `setSelection` to `SetSelection`.

This type seems to be used repeatedly and could be error prone to write out over and over, it seems like there's value in a reusable type.

As far as exporting the type or not, I'm not sure how important the questions is. The truth is that any types that are part of the public interface are effectively public types because they will be visible externally.

This type _is_ directly exported, so I guess that may be the question:

https://github.com/WordPress/gutenberg/blob/1ee2e3c10f60212a9e1e1c27d62d9000401653aa/packages/dataviews/src/index.ts#L4

There, I would tend to be cautious. Is this something that is immediately useful for third parties? If not, then we might still export a single type and use it throughout the repository, but not export it directly. Again, ultimately the type will be visible and therefore accessible externally, the question is really about how much of the public API or how visible it should be. It's not the same to access it in laborious ways or just import the type.

It looks like this type is changing, but it may be strictly wider? All values that satisfied the old type also satisfy the new type? In that case I don't think it would qualify as a breaking change.

One thing that might be interesting around types is proposed by @luisherranz in https://github.com/WordPress/gutenberg/pull/62400. That PR proposes adding automated tests specifically for types. That could be extremely helpful for determining the types of changes in packages because any change that would make code using a previous versions types _invalid_ would be a breaking change.

Reply:
Thanks for the feedback. I capitalised it, made the type private, and used a simpler type.

Reply:
> One thing that might be interesting around types is proposed by @luisherranz in https://github.com/WordPress/gutenberg/pull/62400.

Thank you for mentioning this, @sirreal.

Indeed, I want to propose a way to test types in Gutenberg. This PR shows how it could be done using `tsc`.